src/vs/workbench/services/lifecycle/common/lifecycle.ts

```diff
@@ -65,9 +65,31 @@ export interface BeforeShutdownErrorEvent {
 	readonly error: Error;
 }
 
+export enum WillShutdownJoinerOrder {
+
+	/**
+	 * Joiners to run before the `Last` joiners.
+	 */
+	Default = 1,
+
+	/**
+	 * The joiners to run last.
+	 */
+	Last
+}
+
 export interface IWillShutdownEventJoiner {
-	id: string;
-	label: string;
+	readonly id: string;
+	readonly label: string;
+	readonly order?: WillShutdownJoinerOrder;
+}
+
+export interface IWillShutdownEventDefaultJoiner extends IWillShutdownEventJoiner {
+	readonly order?: WillShutdownJoinerOrder.Default;
+}
+
+export interface IWillShutdownEventLastJoiner extends IWillShutdownEventJoiner {
+	readonly order: WillShutdownJoinerOrder.Last;
 }
 
```

Comment:
If `Last` order is used, we must require a `() => Promise`, otherwise we cannot guarantee last execution. As for the other case, we can support just `Promise` as before.